src/App.tsx

```diff
@@ -15,35 +16,160 @@ export const goodsFromServer = [
   'Garlic',
 ];
 
+export const initialState = {
+  sortedGoods: [...goodsFromServer],
+  buttonSortAplha: { 'is-light': true, 'button is-info': true },
+  buttonSortLength: {
+    'is-light': true,
+    'button is-info is-success': true,
+  },
+  buttonReverse: { button: true, 'is-warning': true, 'is-light': true },
+  buttonRestart: false,
+};
+
+export enum SortType {
+  ALPHA = 'alpha',
+  LENGTH = 'length',
+  REVERSE = 'reverse',
+  DEFAULT = 'default',
+  NONE = 'none',
+}
+
+export function updateState(
+  state: typeof initialState,
+  action: SortType,
+): typeof initialState {
+  switch (action) {
+    case SortType.ALPHA:
+      {
+        const sorted = state.buttonReverse['is-light']
+          ? [...state.sortedGoods].sort((a, b) => a.localeCompare(b))
+          : [...state.sortedGoods].sort((a, b) => b.localeCompare(a));
+
+        return {
+          ...state,
+          sortedGoods: [...sorted],
+          buttonSortAplha: {
+            ...state.buttonSortAplha,
+            'is-light': false,
+          },
+          buttonSortLength: {
+            ...state.buttonSortLength,
+            'is-light': true,
+          },
+          buttonRestart: true,
+        };
+      }
+
+      break;
+    case SortType.LENGTH:
+      {
+        let sorted = [...state.sortedGoods].sort((a, b) => a.length - b.length);
+
+        if (!state.buttonReverse['is-light']) {
+          sorted = [...state.sortedGoods].sort((a, b) => b.length - a.length);
+        }
+
+        return {
+          ...state,
+          sortedGoods: [...sorted],
+          buttonSortAplha: { ...state.buttonSortAplha, 'is-light': true },
+          buttonSortLength: {
+            ...state.buttonSortLength,
+            'is-light': false,
+          },
+          buttonRestart: true,
+        };
+      }
+
+      break;
+    case SortType.REVERSE:
+      {
+        return {
+          ...state,
+          sortedGoods: [...state.sortedGoods.reverse()],
+          buttonReverse: {
+            button: true,
+            'is-warning': true,
+            'is-light': !state.buttonReverse['is-light'],
+          },
+          buttonRestart:
+            !state.buttonSortAplha['is-light'] ||
+            !state.buttonSortLength['is-light'] ||
+            !state.buttonRestart,
+        };
+      }
+
+      break;
+    default:
+      return {
+        ...initialState,
+        sortedGoods: [...goodsFromServer],
+      };
+      break;
+  }
+}
+
 export const App: React.FC = () => {
+  const [action, setAction] = useState(SortType.DEFAULT);
+  const [state, setState] = useState(initialState);
+
+  useEffect(() => {
+    if (SortType.NONE !== action) {
+      setState(updateState(state, action));
+    }
+  }, [action]);
```

Comment:
`state` must be present in `useEffect`'s dependency array; intentionally omitting a reactive variable in the dependency array is discouraged!

updating your `setState` to `setState(last => updateState(last, action))` should fix this.

Also, your state is behaving much like a reducer; maybe check out [React.useReducer](https://react.dev/reference/react/useReducer) (again, way more complicated than it should be for this simple project)

Reply:
Wow, I really liked these observations